### PR TITLE
WebAuthn authenticator guide For Java

### PR DESCRIPTION
**Pull request by robertnicolo-okta**
## Description:
WebAuthn authenticator guide For Java

### Resolves:

* [OKTA-459404](https://oktainc.atlassian.net/browse/OKTA-459404)

### Deploy link to guide (Current as of April 29th, 2022  2:09PM EST)

https://626c24567653de00a8de84b7--reverent-murdock-829d24.netlify.app/docs/guides/authenticators-web-authn/java/main/

### CODE REVIEW

File: packages/@okta/vuepress-site/docs/guides/authenticators-web-authn/main/java/challengeintegrationsteps.md [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+  "clientDataJSON": Binary data ... ,
+  "signature": Binary data ... ,
+  "userHandle": Binary data ... ,
+"type": "public-key"
+}
+```
+
+### 8: Build parameter for sending the signature to Okta
+
+The returned `PublicKeyCredential` object contains the signature and other binary formatted data that you need to convert to strings before sending it back to the Okta servers. The sample app uses client-side javascript to convert this data into a variable.
```
Comment (danielmaharry-okta) [inline packages/@okta/vuepress-site/docs/guides/authenticators-web-authn/main/java/challengeintegrationsteps.md:189]:
```suggestion
The returned `PublicKeyCredential` object contains the signature and other binary-formatted data that you need to convert to strings before sending it back to the Okta servers. The sample app uses client-side javascript to convert this data into a variable.
```

File: packages/@okta/vuepress-site/docs/guides/authenticators-web-authn/main/java/challengeintegrationsteps.md [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+function binToStr(bin) {
+          return btoa(new Uint8Array(bin).reduce((s, byte) => s + String.fromCharCode(byte), ''));
+      }
+```
+
+### 9: Forward signature to Okta for validation
+
+Forward the signature to Okta for validation. Specifically, perform the following steps:
+
+1. First, send the converted signature data to the Okta SDK. The sample app sends the data through an HTTP Post request from the client browser to the server.
```
Comment (danielmaharry-okta) [inline packages/@okta/vuepress-site/docs/guides/authenticators-web-authn/main/java/challengeintegrationsteps.md:218]:
```suggestion
1. Send the converted signature data from the client browser to the server side of your web application.
```

File: packages/@okta/vuepress-site/docs/guides/authenticators-web-authn/main/java/challengeintegrationsteps.md [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+
+    ```javascript
+    fetch('/verify-webauthn', options)
+                  .then(res => { ...
+    ```
+
+2. Next, send the data to `IDXAuthenticationWrapper.verifyWebAuthn()` to have the server validate the signature corresponding to the challenge and public key.
```
Comment (danielmaharry-okta) [inline packages/@okta/vuepress-site/docs/guides/authenticators-web-authn/main/java/challengeintegrationsteps.md:225]:
```suggestion
2. Call `IDXAuthenticationWrapper.verifyWebAuthn()` to have the server validate the signature corresponding to the challenge and public key.
```

File: packages/@okta/vuepress-site/docs/guides/authenticators-web-authn/main/java/challengeintegrationsteps.md [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+
+    ```java
+    ProceedContext proceedContext = Util.getProceedContextFromSession(session);
+
+    AuthenticationResponse authenticationResponse = idxAuthenticationWrapper.verifyWebAuthn(
+      proceedContext, webauthnRequest);
+    ```
+
+3. Depending on the org configuration, the `AuthenticationResponse` from `idxAuthenticationWrapper.verifyWebAuthn()` can return an `authenticationStatus` of `SUCCESS` along with token information, or another status indicating there are additional remediation steps to complete.
```
Comment (danielmaharry-okta) [inline packages/@okta/vuepress-site/docs/guides/authenticators-web-authn/main/java/challengeintegrationsteps.md:234]:
```suggestion
3. Depending on the org configuration, the `AuthenticationResponse` returned by `idxAuthenticationWrapper.verifyWebAuthn()` can contain an `authenticationStatus` of `SUCCESS` along with token information, or another status indicating there are additional remediation steps to complete.
```

**Review by danielmaharry-okta — CHANGES_REQUESTED** [review]
Hi Rob, 
A few general comments
- Decide how you are capitalising `IDXAuthenticationWrapper` and then use it throughout the guide. The instance of the class is called idxAuth.. but the class itself is IDXAuth... As you can't assume the user is going to call the instance of the class the same thing, I'd suggest using IDXAuth...
- The graphics all have drop shadows on them. I thought Chris didnt want this?
- Alt text and titles for images can be richer. I've given an example and leave it to you to complete the rest.
- Mention Thymeleaf once in a note at the top somewhere. You mention it four times. Perhaps in the software versions bit as people will read that to see if their systems are valid.
- Not sure how much I like saying the "The sample app does this" twelve times during the guide. I don't think you need to say the code comes from the sample app at all as long as it is representative of the code needed to do the thing.
- Check for long code lines overflowing to the right and split them onto two lines or more as necessary.

File: packages/@okta/vuepress-site/docs/guides/authenticators-web-authn/main/java/challengeintegrationsteps.md [hunk 5]
```diff
@@ -0,0 +1,234 @@
+### 1: Start challenge flow
+
+The first step is to initiate a use case that requires authentication. This guide uses the sign-in with username and password flow that is initiated with calls to `IDXAuthenticatorWrapper.begin()`, `AuthenticationResponse.getProceedContext()`, and `IDXAuthenticatorWrapper.authenticate()`.
```
[the remark this answers is not in the document]
Comment (robertnicolo-okta) [reply packages/@okta/vuepress-site/docs/guides/authenticators-web-authn/main/java/challengeintegrationsteps.md:3]: updated

File: packages/@okta/vuepress-site/docs/guides/authenticators-web-authn/main/java/challengeintegrationsteps.md [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+ProceedContext proceedContext = beginResponse.getProceedContext();
+
+AuthenticationResponse authenticationResponse;
+authenticationResponse = idxAuthenticationWrapper.authenticate(
+          new AuthenticationOptions(username, password.toCharArray()), proceedContext);
+```
+
+### 2: Display WebAuthn authenticator option
+
+After the user submits their password, make a call to `IDXAuthenticationWrapper.verifyAuthenticator()` to verify their password. If you configure your Okta org as detailed in [Configuration updates](#update-configurations) and WebAuthn is already [enrolled](#integrate-sdk-for-authenticator-enrollment) for the user, `verifyAuthenticator()` returns an `AuthenticationResponse` object with `authenticationStatus` equal to `AWAITING_AUTHENTICATOR_SELECTION` and a `webauthn` authenticator item in the `authenticators` array.
```
[the remark this answers is not in the document]
Comment (robertnicolo-okta) [reply packages/@okta/vuepress-site/docs/guides/authenticators-web-authn/main/java/challengeintegrationsteps.md:16]: updated.

File: packages/@okta/vuepress-site/docs/guides/authenticators-web-authn/main/java/enrollmentintegrationsummary.md [hunk 7]
```diff
@@ -0,0 +1,5 @@
+<div class="common-image-format">
+
+![Sequence diagram showing WebAuthn enrollment summary](/img/authenticators/authenticators-webauthn-java-enrollment-summary.png)
```
[the remark this answers is not in the document]
Comment (robertnicolo-okta) [reply packages/@okta/vuepress-site/docs/guides/authenticators-web-authn/main/java/enrollmentintegrationsummary.md:3]: Alex made the update

File: packages/@okta/vuepress-site/docs/guides/authenticators-web-authn/main/java/challengeintegrationsummary.md [hunk 8]
```diff
@@ -0,0 +1,5 @@
+<div class="common-image-format">
+
+![Sequence diagram showing WebAuthn challenge summary](/img/authenticators/authenticators-webauthn-java-challenge-summary.png)
```
[the remark this answers is not in the document]
Comment (robertnicolo-okta) [reply packages/@okta/vuepress-site/docs/guides/authenticators-web-authn/main/java/challengeintegrationsummary.md:3]: Alex made the update

File: packages/@okta/vuepress-site/docs/guides/authenticators-web-authn/main/java/enrollmentintegrationsteps.md [hunk 9]
```diff
@@ -0,0 +1,244 @@
+### 1: Start enrollment flow
+
+The first step is to initiate a use case that requires authentication. This guide uses the sign-in with username and password flow that is initiated with calls to `IDXAuthenticatorWrapper.begin()`, `AuthenticationResponse.getProceedContext()`, and `IDXAuthenticatorWrapper.authenticate()`.
```
[the remark this answers is not in the document]
Comment (robertnicolo-okta) [reply packages/@okta/vuepress-site/docs/guides/authenticators-web-authn/main/java/enrollmentintegrationsteps.md:3]: updated.

File: packages/@okta/vuepress-site/docs/guides/authenticators-web-authn/main/java/enrollmentintegrationsteps.md [hunk 10]
```diff
@@ -0,0 +1,244 @@
...
+
+```java
+AuthenticationResponse beginResponse = idxAuthenticationWrapper.begin();
```
[the remark this answers is not in the document]
Comment (robertnicolo-okta) [reply packages/@okta/vuepress-site/docs/guides/authenticators-web-authn/main/java/enrollmentintegrationsteps.md:6]: I'm confused over this one. This code block is identical in both flows.

File: packages/@okta/vuepress-site/docs/guides/authenticators-web-authn/main/java/enrollmentintegrationsteps.md [hunk 11]
```diff
@@ -0,0 +1,244 @@
...
+ProceedContext proceedContext = beginResponse.getProceedContext();
+
+AuthenticationResponse authenticationResponse;
+authenticationResponse = idxAuthenticationWrapper.authenticate(
+          new AuthenticationOptions(username, password.toCharArray()), proceedContext);
+```
+
+### 2: Display WebAuthn option
+
+After the user submits their password, make a call to `IDXAuthenticationWrapper.verifyAuthenticator()` to verify their password. If you configure your Okta org as detailed in [Configuration updates](#update-configurations) and WebAuthn is **not** already [enrolled](#integrate-sdk-for-authenticator-enrollment) for the user, `verifyAuthenticator()` returns an `AuthenticationResponse` object with `authenticationStatus` equal to `AWAITING_AUTHENTICATOR_ENROLLMENT_SELECTION` and a `webauthn` authenticator item in the `authenticators` array.
```
[the remark this answers is not in the document]
Comment (robertnicolo-okta) [reply packages/@okta/vuepress-site/docs/guides/authenticators-web-authn/main/java/enrollmentintegrationsteps.md:16]: I changed it to authenticate().

File: packages/@okta/vuepress-site/docs/guides/authenticators-web-authn/main/java/enrollmentintegrationsteps.md [hunk 12]
```diff
@@ -0,0 +1,244 @@
...
+```xml
+<tr th:each="authenticator : ${authenticators}">
+    <div class="form-check">
+        <input class="form-check-input" type="radio" name="authenticator-type" th:value="${authenticator.label}" checked="true">
+        <label class="form-check-label" th:text="${authenticator.label}"></label>
+    </div>
+</tr>
+```
+
+Authenticator selection page example from the sample app:
```
[the remark this answers is not in the document]
Comment (robertnicolo-okta) [reply packages/@okta/vuepress-site/docs/guides/authenticators-web-authn/main/java/enrollmentintegrationsteps.md:47]: updated.

File: packages/@okta/vuepress-site/docs/guides/authenticators-web-authn/main/java/enrollmentintegrationsteps.md [hunk 13]
```diff
@@ -0,0 +1,244 @@
...
+
+</div>
+
+### 3:  Submit WebAuthn option
+
+When the user selects the WebAuthn option and clicks submit, call `IDXAuthenticatorWrapper.enrollAuthenticator()` passing in `ProceedContext` and the authenticator Id returned from `AuthenticationResponse.authenticators[n].factors[n].id`.
+
+```java
+Optional<Authenticator> authenticatorOptional =
+                    authenticators.stream().filter(auth -> auth.getType().equals("webauthn")).findFirst();
```
[the remark this answers is not in the document]
Comment (robertnicolo-okta) [reply packages/@okta/vuepress-site/docs/guides/authenticators-web-authn/main/java/enrollmentintegrationsteps.md:61]: updated.

File: packages/@okta/vuepress-site/docs/guides/authenticators-web-authn/main/java/enrollmentintegrationsteps.md [hunk 14]
```diff
@@ -0,0 +1,244 @@
...
+				}
+			}
+		}
+	}
+}
+```
+
+### 5: Display page to verify WebAuthn credentials
+
+Redirect the user to a page that creates the WebAuthn credentials and allow this page access to the `AuthenticationResponse` properties. Using the sample app as guide, access these properties with the following steps:
```
[the remark this answers is not in the document]
Comment (robertnicolo-okta) [reply packages/@okta/vuepress-site/docs/guides/authenticators-web-authn/main/java/enrollmentintegrationsteps.md:100]: removed.

File: packages/@okta/vuepress-site/docs/guides/authenticators-web-authn/main/java/enrollmentintegrationsteps.md [hunk 15]
```diff
@@ -0,0 +1,244 @@
...
+
+```javascript
+navigator.credentials.create({
+        publicKey: publicKeyCredentialCreationOptions
+    }).then((newCredential) ...
+```
+
+This call initiates the following steps:
+
+1. Browser prompts the user to choose an authenticator.
```
[the remark this answers is not in the document]
Comment (robertnicolo-okta) [reply packages/@okta/vuepress-site/docs/guides/authenticators-web-authn/main/java/enrollmentintegrationsteps.md:169]: updated.

File: packages/@okta/vuepress-site/docs/guides/authenticators-web-authn/main/java/enrollmentintegrationsteps.md [hunk 16]
```diff
@@ -0,0 +1,244 @@
...
+
+<div class="common-image-format">
+
+![UI showing the authenticator options](/img/authenticators/authenticators-webauthn-authenticator-options.png)
+
+</div>
+
+2. After the user chooses the authenticator, the device's local authenticator asks for consent to create the credentials. In the following example, the **Touch ID** authenticator prompts the user for a fingerprint to confirm the consent.
```
[the remark this answers is not in the document]
Comment (robertnicolo-okta) [reply packages/@okta/vuepress-site/docs/guides/authenticators-web-authn/main/java/enrollmentintegrationsteps.md:177]: A local authenticator can be the platform or a roaming authenticator like a yubi key. That content did come from a reputable source but it does sound odd. Without needing to get to low level, I changed "device's local authenticator asks for conset" to just "device asks for consent".

File: packages/@okta/vuepress-site/docs/guides/authenticators-web-authn/main/java/enrollmentintegrationsteps.md [hunk 17]
```diff
@@ -0,0 +1,244 @@
...
+    "response" : {
+      "attestationObject" : Binary data ... ,
+      "clientDataJSON": Binary data ... ,
+    "type": "public-key"
+    }
+    ```
+
+### 8: Build the parameter for sending the public key to the Okta server
+
+The returned `PublicKeyCredential` object contains the signature and other binary formatted data that you need to convert to strings before sending it back to the Okta servers. The sample app uses client-side javascript to convert this data into a variable.
```
[the remark this answers is not in the document]
Comment (robertnicolo-okta) [reply packages/@okta/vuepress-site/docs/guides/authenticators-web-authn/main/java/enrollmentintegrationsteps.md:200]: updated.

File: packages/@okta/vuepress-site/docs/guides/authenticators-web-authn/main/java/enrollmentintegrationsteps.md [hunk 18]
```diff
@@ -0,0 +1,244 @@
...
+
+<div class="common-image-format">
+
+![Page showing webauthn option in list](/img/authenticators/authenticators-webauthn-java-enroll-dropdown-selection.png)
+
+</div>
+
+### 3:  Submit WebAuthn option
+
+When the user selects the WebAuthn option and clicks submit, call `IDXAuthenticatorWrapper.enrollAuthenticator()` passing in `ProceedContext` and the authenticator Id returned from `AuthenticationResponse.authenticators[n].factors[n].id`.
```
[the remark this answers is not in the document]
Comment (robertnicolo-okta) [reply packages/@okta/vuepress-site/docs/guides/authenticators-web-authn/main/java/enrollmentintegrationsteps.md:57]: updated.

File: packages/@okta/vuepress-site/docs/guides/authenticators-web-authn/main/java/enrollmentintegrationsteps.md [hunk 19]
```diff
@@ -0,0 +1,244 @@
...
+2. Next, send the data to `IDXAuthenticationWrapper.verifyWebAuthn()` to have the server validate the signature and store the credential ID and public key associated with the user's account.
+
+    ```java
+    ProceedContext proceedContext = Util.getProceedContextFromSession(session);
+
+    AuthenticationResponse authenticationResponse = idxAuthenticationWrapper.verifyWebAuthn(
+      proceedContext, webauthnRequest);
+    ```
+
+3. Depending on the org configuration, the `AuthenticationResponse` from `idxAuthenticationWrapper.verifyWebAuthn()` can return an `authenticationStatus` of `SUCCESS` along with token information, or another status indicating there are additional remediation steps to complete.
```
[the remark this answers is not in the document]
Comment (robertnicolo-okta) [reply packages/@okta/vuepress-site/docs/guides/authenticators-web-authn/main/java/enrollmentintegrationsteps.md:244]: updated.

File: packages/@okta/vuepress-site/docs/guides/authenticators-web-authn/main/java/challengeintegrationsteps.md [hunk 20]
```diff
@@ -0,0 +1,234 @@
...
+
+<div class="common-image-format">
+
+![Page showing webauthn option in list](/img/authenticators/authenticators-webauthn-java-dropdown-selection.png)
+
+</div>
+
+### 3: Submit WebAuthn authenticator option
+
+When the user selects the WebAuthn option and clicks submit, call `IDXAuthenticatorWrapper.enrollAuthenticator()` passing in `ProceedContext` and the authenticator Id returned from `AuthenticationResponse.authenticators[n].factors[n].id`.
```
[the remark this answers is not in the document]
Comment (robertnicolo-okta) [reply packages/@okta/vuepress-site/docs/guides/authenticators-web-authn/main/java/challengeintegrationsteps.md:56]: updated.

[the remark this answers is not in the document]
Comment (robertnicolo-okta) [reply packages/@okta/vuepress-site/docs/guides/authenticators-web-authn/main/java/challengeintegrationsteps.md:56]: updated.

File: packages/@okta/vuepress-site/docs/guides/authenticators-web-authn/main/java/challengeintegrationsteps.md [hunk 21]
```diff
@@ -0,0 +1,234 @@
...
+                    authenticators.stream().filter(auth -> auth.getType().equals("webauthn")).findFirst();
+String authId = authenticatorOptional.get().getId();
+
+AuthenticationResponse enrollResponse = idxAuthenticationWrapper.enrollAuthenticator(proceedContext, authId);
+
+```
+
+### 4: Identify data for getting the credential
+
+The `AuthenticationResponse` object from `IDXAuthenticatorWrapper.enrollAuthenticator()` has `authenticationStatus` set to `AWAITING_AUTHENTICATOR_VERIFICATION`, which indicates the user must verify their WebAuthn credentials. The code below shows an example response. Additionally, `AuthenticationResponse` returns the challenge, credential Id, and other information needed to verify the WebAuthn credentials on the user's device.
```
[the remark this answers is not in the document]
Comment (robertnicolo-okta) [reply packages/@okta/vuepress-site/docs/guides/authenticators-web-authn/main/java/challengeintegrationsteps.md:69]: Updated.

File: packages/@okta/vuepress-site/docs/guides/authenticators-web-authn/main/java/challengeintegrationsteps.md [hunk 22]
```diff
@@ -0,0 +1,234 @@
...
+      }
+    },
+    "webauthnCredentialId": "AYqpxcR9Jrw6BzVJyZf-ImP7OffDl9-pHcRV2fLD9wexskXac7-DFXrX29A7oFURAwT1oFKoI1loau..."
+  }
+}
+```
+
+### 5: Display page to verify WebAuthn credentials
+
+Redirect the user to a page that verifies the WebAuthn credentials. Allow this page access to `AuthenticationResponse.webAuthnParams.currentAuthenticator.contextualData.challengeData` and  `AuthenticationResponse.webAuthnParams.webauthnCredentialId`. Using the sample app as guide, access these properties with the following steps:
```
[the remark this answers is not in the document]
Comment (robertnicolo-okta) [reply packages/@okta/vuepress-site/docs/guides/authenticators-web-authn/main/java/challengeintegrationsteps.md:99]: updated.

File: packages/@okta/vuepress-site/docs/guides/authenticators-web-authn/main/java/challengeintegrationsteps.md [hunk 23]
```diff
@@ -0,0 +1,234 @@
...
+    ```java
+    String webauthnCredentialId = enrollResponse.getWebAuthnParams().getWebauthnCredentialId();
+
+    modelAndView.addObject("title", "Select Webauthn Authenticator");
+    modelAndView.addObject("webauthnCredentialId", webauthnCredentialId);
+    modelAndView.addObject("challengeData", enrollResponse.getWebAuthnParams()
+            .getCurrentAuthenticator().getValue().getContextualData().getChallengeData());
+    ```
+
+2. Set client-side javascript variables. The sample app uses the [Thymeleaf](https://www.thymeleaf.org/) template engine to set these variables.
```
[the remark this answers is not in the document]
Comment (robertnicolo-okta) [reply packages/@okta/vuepress-site/docs/guides/authenticators-web-authn/main/java/challengeintegrationsteps.md:112]: They are used to either get or create the credential. I've updated the text to give more context.

File: packages/@okta/vuepress-site/docs/guides/authenticators-web-authn/main/java/challengeintegrationsteps.md [hunk 24]
```diff
@@ -0,0 +1,234 @@
...
+
+    ```javascript
+    <script th:inline="javascript">
+        const challengeData = /*[[${challengeData}]]*/ '';
+        const webauthnCredentialId = /*[[${webauthnCredentialId}]]*/ '';
+    </script>
+    ```
+
+    The following example renders the previous javascript code snippet.
```
[the remark this answers is not in the document]
Comment (robertnicolo-okta) [reply packages/@okta/vuepress-site/docs/guides/authenticators-web-authn/main/java/challengeintegrationsteps.md:121]: Updated.

File: packages/@okta/vuepress-site/docs/guides/authenticators-web-authn/main/java/challengeintegrationsteps.md [hunk 25]
```diff
@@ -0,0 +1,234 @@
...
+  };
+
+function strToBin(str) {
+    return Uint8Array.from(atob(base64UrlSafeToBase64(str)), c => c.charCodeAt(0));
+}
+```
+
+### 7: Receive credential and create cryptographic signature
+
+Call the Web Authentication API's `navigator.credentials.get()` in the client browser and pass in the `CredentialRequestOptions` object created in the previous step.
```
[the remark this answers is not in the document]
Comment (robertnicolo-okta) [reply packages/@okta/vuepress-site/docs/guides/authenticators-web-authn/main/java/challengeintegrationsteps.md:152]: updated.

File: packages/@okta/vuepress-site/docs/guides/authenticators-web-authn/main/java/challengeintegrationsteps.md (hunk 1 above)
Comment (robertnicolo-okta) [reply packages/@okta/vuepress-site/docs/guides/authenticators-web-authn/main/java/challengeintegrationsteps.md:189]: updated.

File: packages/@okta/vuepress-site/docs/guides/authenticators-web-authn/main/java/challengeintegrationsteps.md (hunk 3 above)
Comment (robertnicolo-okta) [reply packages/@okta/vuepress-site/docs/guides/authenticators-web-authn/main/java/challengeintegrationsteps.md:225]: Updated.

File: packages/@okta/vuepress-site/docs/guides/authenticators-web-authn/main/java/challengeintegrationsteps.md (hunk 4 above)
Comment (robertnicolo-okta) [reply packages/@okta/vuepress-site/docs/guides/authenticators-web-authn/main/java/challengeintegrationsteps.md:234]: Updated.

File: packages/@okta/vuepress-site/docs/guides/authenticators-web-authn/main/java/enrollmentintegrationsteps.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (robertnicolo-okta) [reply packages/@okta/vuepress-site/docs/guides/authenticators-web-authn/main/java/enrollmentintegrationsteps.md]: I confirmed with Arvind. The Java sample app can either be identifier first or require username and password based on an org setting. In this guide we are discussing the username and pwd flow.

**Review by danielmaharry-okta — APPROVED** [review]
LGTM!

File: packages/@okta/vuepress-site/docs/guides/authenticators-web-authn/main/java/challengeintegrationsteps.md [hunk 26]
```diff
@@ -0,0 +1,235 @@
+### 1: Start challenge flow
+
+Start the challenge flow with calls to `IDXAuthenticatorWrapper.begin()` and `AuthenticationResponse.getProceedContext()`. Then send username and password to the Okta server with `IDXAuthenticatorWrapper.authenticate()`.
```
Comment (susanharper-okta) [inline packages/@okta/vuepress-site/docs/guides/authenticators-web-authn/main/java/challengeintegrationsteps.md:3]:
```suggestion
Start the challenge flow with calls to `IDXAuthenticatorWrapper.begin()` and `AuthenticationResponse.getProceedContext()`. Then send the username and password to the Okta server with `IDXAuthenticatorWrapper.authenticate()`.
```

File: packages/@okta/vuepress-site/docs/guides/authenticators-web-authn/main/java/challengeintegrationsteps.md [hunk 27]
```diff
@@ -0,0 +1,235 @@
...
+2. Call `IDXAuthenticationWrapper.verifyWebAuthn()` to have the server validate the signature corresponding to the challenge and public key.
+
+    ```java
+    ProceedContext proceedContext = Util.getProceedContextFromSession(session);
+
+    AuthenticationResponse authenticationResponse = idxAuthenticationWrapper.verifyWebAuthn(
+      proceedContext, webauthnRequest);
+    ```
+
+3. Depending on the org configuration, the `AuthenticationResponse` returned by `IDXAuthenticationWrapper.verifyWebAuthn()` may contain `authenticationStatus` of `SUCCESS` along with token information, or another status indicating there are further remediation steps to complete.
```
Comment (susanharper-okta) [inline packages/@okta/vuepress-site/docs/guides/authenticators-web-authn/main/java/challengeintegrationsteps.md:235]:
```suggestion
3. Depending on the org configuration, the `AuthenticationResponse` returned by `IDXAuthenticationWrapper.verifyWebAuthn()` may contain `authenticationStatus` of `SUCCESS` along with token information or another status to indicate that there are further remediation steps to complete.
```

File: packages/@okta/vuepress-site/docs/guides/authenticators-web-authn/main/java/enrollmentintegrationsteps.md [hunk 28]
```diff
@@ -0,0 +1,246 @@
...
+String authId = authenticatorOptional.get().getId();
+
+AuthenticationResponse enrollResponse =
+  idxAuthenticationWrapper.enrollAuthenticator(proceedContext, authId);
+
+```
+
+### 4: Identify data for creating a new credential
+
+The `AuthenticationResponse` object returned by `IDXAuthenticatorWrapper.enrollAuthenticator()` has `authenticationStatus` set to `AWAITING_AUTHENTICATOR_VERIFICATION`, indicating the user must verify their WebAuthn credentials. `AuthenticationResponse` returns the challenge and other information needed to verify the WebAuthn credentials on the user's device.
```
Comment (susanharper-okta) [inline packages/@okta/vuepress-site/docs/guides/authenticators-web-authn/main/java/enrollmentintegrationsteps.md:72]:
```suggestion
The `AuthenticationResponse` object returned by `IDXAuthenticatorWrapper.enrollAuthenticator()` has `authenticationStatus` set to `AWAITING_AUTHENTICATOR_VERIFICATION`, which indicates that the user must verify their WebAuthn credentials. `AuthenticationResponse` returns the challenge and other information needed to verify the WebAuthn credentials on the user's device.
```

File: packages/@okta/vuepress-site/docs/guides/authenticators-web-authn/main/java/enrollmentintegrationsteps.md [hunk 29]
```diff
@@ -0,0 +1,246 @@
...
+1. Call `ModelandView.addObject()` and add `AuthenticationResponse.currentAuthenticator`, which is used later to extract the `challenge` and other data.
+
+    ```java
+    modelAndView = new ModelAndView("enroll-webauthn-authenticator");
+    modelAndView.addObject("title", "Enroll WebAuthn Authenticator");
+    modelAndView.addObject("currentAuthenticator",
+            enrollResponse.getWebAuthnParams().getCurrentAuthenticator());
+    ```
+
+2. Set client-side javascript variables to values originating from `currentAuthenticator`. These variables are used later to create the credential.
```
Comment (susanharper-okta) [inline packages/@okta/vuepress-site/docs/guides/authenticators-web-authn/main/java/enrollmentintegrationsteps.md:113]:
```suggestion
2. Set client-side javascript variables to values that originate from `currentAuthenticator`. These variables are used later to create the credential.
```

File: packages/@okta/vuepress-site/docs/guides/authenticators-web-authn/main/java/enrollmentintegrationsteps.md [hunk 30]
```diff
@@ -0,0 +1,246 @@
...
+
+```javascript
+navigator.credentials.create({
+        publicKey: publicKeyCredentialCreationOptions
+    }).then((newCredential) ...
+```
+
+This call initiates the following steps:
+
+1. The browser prompts the user to choose the type of WebAuthn authenticator they are using. For example, a portable hardware authenticator such as a USB security key, or a software-based authenticator such as a fingerprint scanner.
```
Comment (susanharper-okta) [inline packages/@okta/vuepress-site/docs/guides/authenticators-web-authn/main/java/enrollmentintegrationsteps.md:171]:
```suggestion
1. The browser prompts the user to choose the type of WebAuthn authenticator that they are using. For example: a portable hardware authenticator such as a USB security key or a software-based authenticator such as a fingerprint scanner
```

File: packages/@okta/vuepress-site/docs/guides/authenticators-web-authn/main/java/enrollmentintegrationsteps.md [hunk 31]
```diff
@@ -0,0 +1,246 @@
...
+
+2. After the user chooses the authenticator, the device asks for consent to create the credentials. In the following example, the **Touch ID** authenticator prompts the user for a fingerprint to confirm the consent.
+
+<div class="common-image-format">
+
+![A prompt from Google Chrome asking to either use Touch ID or password to verify the user's identity](/img/authenticators/authenticators-webauthn-user-consent.png "Verify your identity prompt")
+
+</div></br>
+
+3. The private and public key pairs are created. The private key is stored internally on the device and linked to the user and domain name. Specifically, `navigator.credentials.create()` returns an object of type `PublicKeyCredential` containing the public key, credential ID, and other information used to associate the new credential with the server and browser.
```
Comment (susanharper-okta) [inline packages/@okta/vuepress-site/docs/guides/authenticators-web-authn/main/java/enrollmentintegrationsteps.md:187]:
```suggestion
3. The private and public key pairs are created. The private key is stored internally on the device and linked to the user and domain name. Specifically, `navigator.credentials.create()` returns an object of type `PublicKeyCredential` that contains the public key, credential ID, and other information used to associate the new credential with the server and browser.
```

File: packages/@okta/vuepress-site/docs/guides/authenticators-web-authn/main/java/enrollmentintegrationsteps.md [hunk 32]
```diff
@@ -0,0 +1,246 @@
...
+    "response" : {
+      "attestationObject" : Binary data ... ,
+      "clientDataJSON": Binary data ... ,
+    "type": "public-key"
+    }
+    ```
+
+### 8: Build the parameter for sending the public key to the Okta server
+
+The returned `PublicKeyCredential` object contains the signature and other binary-formatted data you need to convert to strings before sending it back to the Okta servers.
```
Comment (susanharper-okta) [inline packages/@okta/vuepress-site/docs/guides/authenticators-web-authn/main/java/enrollmentintegrationsteps.md:202]:
```suggestion
The returned `PublicKeyCredential` object contains the signature and other binary-formatted data that you need to convert to strings before sending it back to the Okta servers.
```

File: packages/@okta/vuepress-site/docs/guides/authenticators-web-authn/main/java/enrollmentintegrationsteps.md [hunk 33]
```diff
@@ -0,0 +1,246 @@
...
+2. Next, send the data to `IDXAuthenticationWrapper.verifyWebAuthn()` to have the server validate the signature and store the credential ID and public key associated with the user's account.
+
+    ```java
+    ProceedContext proceedContext = Util.getProceedContextFromSession(session);
+
+    AuthenticationResponse authenticationResponse = idxAuthenticationWrapper.verifyWebAuthn(
+      proceedContext, webauthnRequest);
+    ```
+
+3. Depending on the org configuration, the `AuthenticationResponse` returned by `IDXAuthenticationWrapper.verifyWebAuthn()` may contain `authenticationStatus` of `SUCCESS` along with token information, or another status indicating there are further remediation steps to complete.
```
Comment (susanharper-okta) [inline packages/@okta/vuepress-site/docs/guides/authenticators-web-authn/main/java/enrollmentintegrationsteps.md:246]:
```suggestion
3. Depending on the org configuration, the `AuthenticationResponse` returned by `IDXAuthenticationWrapper.verifyWebAuthn()` may contain `authenticationStatus` of `SUCCESS` along with token information, or another status to indicate that there are further remediation steps to complete.
```

File: packages/@okta/vuepress-site/docs/guides/authenticators-web-authn/main/java/orgconfigurepwdonly.md [hunk 34]
```diff
@@ -0,0 +1 @@
+[An Okta org already configured for a password-only use case](/docs/guides/oie-embedded-common-org-setup/java/main/#set-up-your-okta-org-for-a-password-factor-only-use-case)
```
Comment (susanharper-okta) [inline packages/@okta/vuepress-site/docs/guides/authenticators-web-authn/main/java/orgconfigurepwdonly.md:1]:
```suggestion
[An Okta org already configured for a password only use case](/docs/guides/oie-embedded-common-org-setup/java/main/#set-up-your-okta-org-for-a-password-factor-only-use-case)
```

File: packages/@okta/vuepress-site/docs/guides/authenticators-web-authn/main/java/softwareversions.md [hunk 35]
```diff
@@ -0,0 +1,2 @@
+* Your browser must be up-to-date to use WebAuthn. The Embedded SDK sample app for the Java SDK only supports Chrome (version 88 and higher).
```
Comment (susanharper-okta) [inline packages/@okta/vuepress-site/docs/guides/authenticators-web-authn/main/java/softwareversions.md:1]:
```suggestion
* Your browser must be up to date to use WebAuthn. The Embedded SDK sample app for the Java SDK only supports Chrome (version 88 and higher).
```

**Review by susanharper-okta — COMMENTED** [review]
couple things. additionally, some comments about text in the graphics:

authenticators-webauthn-java-challenge-summary.png

id -> ID

Step 4: (I think?) Call enrollAuthenticator passing in authenticator ID

Step 6: (I think?) Set up age with challenge and credential ID

Verify that

Step 7: Verify that credential ID matches domain name

avoid e.g. - use example: and finger print -> fingerprint

Step 7: User initiates verification (example: fingerprint)

-------------
authenticators-webauthn-java-enrollment-summary.png

id -> ID

Step 4: (I think) Call enrollAutenticator passing in authenticator ID

avoid e.g. - use example:, Id -> ID, and finger print -> fingerprint

Step 7: User chooses authenticator (example: Touch ID)

Step 7: User initiates verification (example: fingerprint)

File: packages/@okta/vuepress-site/docs/guides/authenticators-web-authn/main/java/challengeintegrationsteps.md (hunk 26 above)
Comment (robertnicolo-okta) [reply packages/@okta/vuepress-site/docs/guides/authenticators-web-authn/main/java/challengeintegrationsteps.md]: updated

**Pull request merged**